planner/core/rule_partition_processor.go

```diff
@@ -190,14 +190,24 @@ func (s *partitionProcessor) findUsedPartitions(ctx sessionctx.Context, tbl tabl
 						}
 						used = append(used, int(idx))
 					}
-				} else {
-					used = []int{FullRange}
-					break
+					continue
+				}
+
+				// maximum number of partitions is 1024
+				if col.RetType.Flen < 10 {
+					// all possible hash values
+					maxUsedPartitions := 1 << col.RetType.Flen
+					if maxUsedPartitions < numPartitions {
+						for i := 0; i < maxUsedPartitions; i++ {
```

Comment:
Find a bad case of this prunning rule:
```
mysql> create table te(a int(2)) partition by hash(a) partitions 30;
Query OK, 0 rows affected (0.01 sec)

mysql> insert into te values (111);
Query OK, 1 row affected (0.00 sec)

mysql> select * from te;
Empty set (0.00 sec)
```

Reply:
This is because `Flen` doesn't reflect the length of real storage bits if the type is `INT`. This should be fixed in the new commit.